### PR TITLE
Inline Hooks Documentation

### PR DESCRIPTION
**Pull request by motikagan-okta**
## Description:
- Documentation for Inline Hooks.
- Should not be merged until Inline Hook framework and API AM Inline Hook are released.
- Currently planned for 2019.02.0 release.

### Resolves:
* [OKTA-198337](https://oktainc.atlassian.net/browse/OKTA-198337)
* [OKTA-197050](https://oktainc.atlassian.net/browse/OKTA-197050)
* [OKTA-201582](https://oktainc.atlassian.net/browse/OKTA-201582)

### CODE REVIEW

File: _source/_docs/api/resources/inline-hooks.md [hunk 1]
```diff
@@ -0,0 +1,614 @@
...
+-H "Authorization: SSWS ${api_token}" \
+-d '{
+    "name" : "My Test Inline Hook",
+    "type" : "com.okta.tokens.transform",
+    "version" : "1.0.0",
+    "channel" : {
+        "type" : "HTTP",
+        "version" : "1.0.0",
+        "config" : {
+            "uri" : "http://127.0.0.1:4567/inlineHook",
```
Comment (amoghvasekar-okta) [inline _source/_docs/api/resources/inline-hooks.md]: nit : can we use `https` everywhere in the docs, since that is a requirement for a URL?

File: _source/_docs/api/resources/inline-hooks.md [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+            "method" : "POST",
```
Comment (amoghvasekar-okta) [inline _source/_docs/api/resources/inline-hooks.md]: minor : You can omit the `method` for now, since we default (and currently should document) only `POST` method.

File: _source/_docs/api/resources/inline-hooks.md [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+                {
+                    "key": "X-Other-Header",
+                    "value": "some-other-value"
+                }
+            ],
+            "method": "POST",
+            "authScheme": {
+                "type": "HEADER",
+                "key": "Authorization",
+                "value": "******"
```
Comment (amoghvasekar-okta) [inline _source/_docs/api/resources/inline-hooks.md]: After API design review we decided to omit sending back the `value` instead of masking it, as the `value` is essentially a secret as sensitive as a password. Thus, the `value` is a WRITE-ONLY property.
It would be good to document this fact, and omit the value in examples. See https://developer.okta.com/docs/api/resources/users#password-object for an example.

File: _source/_docs/api/resources/inline-hooks.md [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+| status      | Status of the Inline Hook, INACTIVE will block execution. | Enum           | FALSE    | FALSE  | FALSE    |           |           | One of ACTIVE, INACTIVE                           |
+| name        | Display name for Inline Hook.                             | String         | FALSE    | TRUE   | FALSE    | 1         | 255       |                                                   |
+| type        | Type of the Inline Hook.                                  | inlineHookType | FALSE    | FALSE  | TRUE     |           |           | Immutable after Inline Hook creation              |
+| version     | Version of the Channel.                                   | Integer        | FALSE    | FALSE  | TRUE     |           |           | Must match a valid version number                 |
+| channel     | Channel for the Inline Hook.                              | Channel        | FALSE    | FALSE  | FALSE    |           |           | Validation is determined by the specific Channel. |
```
Comment (amoghvasekar-okta) [inline _source/_docs/api/resources/inline-hooks.md]: Would there be a model explaining the `Channel` object, and fields inside it such as headers and AuthScheme?

File: _source/_use_cases/inline_hooks/index.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+1. Okta receives the response, acts on any commands it includes, and resumes the process flow that originally triggered the inline hook.
+
+### Request and Response Overview
+
+Okta's request to your external service consists of an HTTPS POST request with a JSON payload. The objects included in the JSON payload provide data relevant to the process flow that triggered the inline hook. The set of objects varies depending the type of inline hook you're using.
+
+{% img hook-request-response.png "Hook Request and Response" alt:"Hook Request and Response" %}
+
+Your service needs to handle the inline hook by responding to Okta's request. The JSON payload of the response your service sends can contain a `commands` object, in which you send commands to Okta that affect the course of the Okta process flow. The commands available vary depending on the type of inline hook you're using.
```
Comment (amoghvasekar-okta) [inline _source/_use_cases/inline_hooks/index.md]: Couple of small comments - 
1. Thoughts on including the timeout and retry value, similar to how you documented for the `execute` API method above?
2. Mention the fact that if the customer's service returns a 400 error, the request is not retried? Either here or in the `HTTP Status Code` section below

File: _source/_docs/api/resources/inline-hooks.md [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+{:.api .api-request .api-request-example}
+
+~~~sh
+curl -v -X POST \
+-H "Accept: application/json" \
+-H "Content-Type: application/json" \
+-H "Authorization: SSWS ${api_token}" \
+-d '{
+    "name" : "My Test Inline Hook",
+    "type" : "com.okta.tokens.transform",
```
Comment (amoghvasekar-okta) [inline _source/_docs/api/resources/inline-hooks.md]: @danielzeller-okta - Incorrect type?

File: _source/_docs/api/resources/inline-hooks.md [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+                }
+            ],
+            "authScheme" : {
+                "type" : "HEADER",
+                "key" : "Authorization",
+                "value" : "api-key"
+            }
+        }
+    }
+}' "https://{yourOktaDomain}/api/v1/inlineHook"
```
Comment (amoghvasekar-okta) [inline _source/_docs/api/resources/inline-hooks.md]: typo - inlineHook`s`

File: _source/_docs/api/resources/inline-hooks.md (hunk 6 above)
Comment (danielzeller-okta) [reply _source/_docs/api/resources/inline-hooks.md]: yes- for oauth transforms the type is `com.okta.oauth2.tokens.transform`

Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Updated.

File: _source/_docs/api/resources/inline-hooks.md (hunk 2 above)
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Removed `method`.

File: _source/_docs/api/resources/inline-hooks.md [hunk 8]
```diff
@@ -0,0 +1,614 @@
...
+        }
+    },
+    "created": "2018-05-15T01:23:08.000Z",
+    "lastUpdated": "2018-05-15T01:23:08.000Z"
+}
+~~~
+
+### Get Inline Hook
+
+{% api_operation get /api/v1/inlineHoooks/${inlineHookId}%}
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Fixed typo.

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+The `commands` object is where you can provide commands to Okta. It is where you can tell Okta to add additional claims to the token.
+
+The `commands` object is an array, allowing you to send multiple commands. In each array element, there needs to be a `type` property and `value` property. The `type` property is where you specify which of the supported commands you wish to execute, and `value` is where you supply an operand for that command.
+
+In the case of the API Access Management hook type, the `value` property is itself a nested object, in which you specify a particular operation, a path to act on, and a value.
+
+| Property | Description                                                              | Data Type       |
+|----------|--------------------------------------------------------------------------|-----------------|
+| type     | One of the [supported commands](#supported-commands).                    | String          |
+| value    | Operand to pass to the command. It specifies a particular op to perform. | [value](#value) |
```
Comment (bellatumaneng-okta) [inline _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: `value` is actually an array as well, so you can specify multiple operations for a given command `type`. So your `commands` object could also be:

```
[
  {
    "type": "com.okta.tokens.id_token.patch",
    "value": [
      {
        "op": "add",
        "path": "/claims/patientId",
        "value": 1234
      },
      {
        "op": "add",
        "path": "/claims/otherClaim",
        "value": "second claim value"
      }
    ]
  },
  {
    "type": "com.okta.tokens.access_token.patch",
    "value": [
      {
        "op": "add",
        "path": "/claims/external_guid",
        "value": "F0384685-F87D-474B-848D-2058AC5655A7"
      }
    ]
  }
]
```

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+
+| Property | Description                                                                                                                                                                                                       | Data Type |
+|----------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------|
+| op       | The name of one of the [supported ops](#list-of-supported-ops).                                                                                                                                                   | String    |
+| path     | Location within the token to apply the operation, specified as a slash-delimited path. When adding a claim, this will always begin with `/claims/`,  and be followed by the name of the new claim you are adding. | String    |
+| value    | Value to set the claim to.                                                                                                                                                                                        | Number    |
```
Comment (bellatumaneng-okta) [inline _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: The data type is not necessarily a number – any JSON object should work (e.g. Strings and Arrays would work fine too)

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+```JSON
+{"commands":
+[{
+    "type": "com.okta.identity.patch",
+    "value":
+    [
+        {
+        "op": "add",
+        "path": "/claims/extPatientId",
+        "value": "1234"
```
Comment (bellatumaneng-okta) [inline _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: value is a string instead of a number in this example

File: _source/_docs/api/resources/inline-hooks.md [hunk 12]
```diff
@@ -0,0 +1,583 @@
...
+    "id": "cali2j192cIE9VpHn0h7",
+    "status": "ACTIVE",
+    "name": "Test Inline Hook",
+    "type": "com.okta.oauth2.tokens.transform",
+    "version": "1.0.0",
+    "channel": {
+        "type": "HTTPS",
+        "version": "1.0.0",
+        "config": {
+            "uri": "https://my.test.url",
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Thanks, fixed to use the same example URI.

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+
+## Objects in Response You Send
+
+For the API Access Management Inline hook, the `commands` and `error` objects that you can return in the JSON payload of your response are defined as follows:
+
+### commands
+
+The `commands` object is where you can provide commands to Okta. It is where you can tell Okta to add additional claims to the token.
+
+The `commands` object is an array, allowing you to send multiple commands. In each array element, there needs to be a `type` property and `value` property. The `type` property is where you specify which of the supported commands you wish to execute, and `value` is where you supply an operand for that command.
```
Comment (bellatumaneng-okta) [inline _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: One important detail that we might want to include: the service should only respond with commands that can be applied to the requested tokens, i.e., if the requester is only requesting an ID token, the `commands` array should not contain any commands of the type `com.okta.access.patch`.

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md (hunk 10 above)
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: Corrected data type to be any JSON object.

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md (hunk 13 above)
Comment (bellatumaneng-okta) [reply _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: @johngronberg-okta do we want to specify this?^

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md (hunk 11 above)
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: Have now updated data type for `value` to "any JSON object", so this example should be okay.

File: _source/_docs/api/resources/inline-hooks.md [hunk 14]
```diff
@@ -0,0 +1,583 @@
...
+            "path": "/claims/extPatientId",
+            "value": "1234"
+          }
+        ]
+      },
+      {
+      "type": "com.okta.tokens.access_token.patch",
+      "value":
+
+
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Thanks, removed the extra whitespace.

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md (hunk 9 above)
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: Thanks, Bella. I hadn't noticed that. I've updated the description of the `value` object to say: "It is an array, allowing you to request more than one operation."

File: _source/_docs/api/resources/inline-hooks.md (hunk 7 above)
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Thanks, corrected.

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md (hunk 13 above)
Comment (johngronberg-okta) [reply _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: @bellatumaneng-okta @motikagan-okta yeah I think we should add this detail.

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+
+This type of inline hook is triggered when OAuth 2.0 and OpenID Connect tokens are minted by your Okta Custom Authorization Server. Before sending the token to the requester, Okta calls out to your external service, and your service can respond with commands to add custom claims to the token.
+
+This functionality can be used to add sensitive user data to tokens without having to store that data in Okta user profiles. For example, tokens minted for a medical app can be augmented with confidential patient data provided by your external service and not stored in Okta.
+
+This inline hook works only when using an Okta Custom Authorization Server, not the built-in Okta Authorization Server.
+
+You cannot use this inline hook to overwrite claims in tokens, only to add new ones.
+
+## Objects in the Request from Okta
```
Comment (yuliu-okta) [inline _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: Do we need to doc the objects in the context somewhere? Also do we need to give more detail when Okta will call external service and different protocol examples for different OAuth2 flows? for example, for implicit flow and code flow, Okta calls at the authorize endpoint; for other flows, Okta calls at the token endpoint.

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+
+#### lifetime
+
+| Property   | Description                                                            | Data Type |
+|------------|------------------------------------------------------------------------|-----------|
+| expiration | The expiration time of the token in seconds since January 1, 1970 UTC. | Number    |
+
+#### scopes
+
+The set of scopes that have been granted. See [Tokens and Claims](/docs/api/resources/oidc#tokens-and-claims), in the documentation for Okta's OpenID Connect and OAuth 2.0 API, for descriptions of the scopes that can be included.
```
Comment (jakubvul) [inline _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: I'd rephrase this, like I suggested above ^

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+
+When you return an error object, it should have the following structure:
+
+| Property     | Description                          | Data Type                   |
+|--------------|--------------------------------------|-----------------------------|
+| errorSummary | Human-readable summary of the error. | String                      |
+
+Returning an error object will cause Okta to return an OAuth 2.0 error to the requester of the token, with the value of `error` set to `server_error`, and the value of`error_description` set to the string you supplied in the `errorSummary` property of the `error` object you returned.
+
+## Sample Listing of JSON Payload of Request
```
Comment (jakubvul) [inline _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: This is a very confusing title. What does it actually mean?

File: _source/_docs/api/resources/inline-hooks.md [hunk 18]
```diff
@@ -0,0 +1,581 @@
+---
+layout: docs_page
+title: Inline Hooks
+category: management
+excerpt: The Inline Hooks APIs provides a CRUD and execution interface for Inline Hooks.
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Corrected.

File: _source/_docs/api/resources/inline-hooks.md (hunk 3 above)
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Removed 'channel.authScheme.value' property from response listings. Added note: "The 'channel.authScheme.value' property is write-only. You set it, but it is never returned in the response."

File: _source/_docs/api/resources/inline-hooks.md [hunk 19]
```diff
@@ -0,0 +1,581 @@
...
+}
+~~~
+
+### List Inline Hooks
+
+{% api_operation get /api/v1/inlineHooks%}
+
+| Parameter | Description                    | Param Type | DataType | Required |
+|-----------|--------------------------------|------------|----------|----------|
+| type      | A valid `inlineHookType` name. | Query      | Enum     | FALSE    |
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Corrected.

File: _source/_docs/api/resources/inline-hooks.md [hunk 20]
```diff
@@ -0,0 +1,581 @@
...
+### Update Inline Hook
+
+{% api_operation put /api/v1/inlineHooks/{inlineHookId}%}
+
+#### Request Parameters
+
+| Parameter    | Description             | Param Type | DataType          | Required |
+|--------------|-------------------------|------------|-------------------|----------|
+| inlineHookId | A valid Inline Hook ID. | Path       | String            | TRUE     |
+| inlineHook   | A valid Inline Hook.    | Body       | Inline Hook Model | TRUE     |
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Corrected.

File: _source/_docs/api/resources/inline-hooks.md [hunk 21]
```diff
@@ -0,0 +1,581 @@
...
+
+The submitted Inline Hook will replace the existing version after passing validation. Refer to the above models to see
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Corrected.

File: _source/_docs/api/resources/inline-hooks.md [hunk 22]
```diff
@@ -0,0 +1,581 @@
...
+        }
+    },
+    "created": "2018-05-15T01:23:08.000Z",
+    "lastUpdated": "2018-05-15T01:23:08.000Z"
+}
+~~~
+
+### Delete Inline Hook
+
+{% api_operation delete /api/v1/inlineHooks/{inelinHookId}%}
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Corrected typo. Corrected ${} throughout.

File: _source/_docs/api/resources/inline-hooks.md [hunk 23]
```diff
@@ -0,0 +1,581 @@
...
+
+#### Request Parameters
+
+| Parameter    | Description            | Param Type | DataType | Required |
+|--------------|------------------------|------------|----------|----------|
+| inlineHookId | A valid Inline Hook id | Path       | String   | TRUE     |
+
+Deletes the Inline Hook matching the provided `inlineHook`. Once deleted, this Inline Hook will be unrecoverable.
+As a safety precaution, only Inline Hooks with a status of INACTIVE are eligible for deletion.
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Corrected.

File: _source/_docs/api/resources/inline-hooks.md [hunk 24]
```diff
@@ -0,0 +1,581 @@
...
+    }
+}
+~~~
+
+### Inline Hook Object
+
+| Property    | Description                                               | DataType       | Nullable | Unique | ReadOnly | MinLength | MaxLength | Validation                                        |
+|-------------|-----------------------------------------------------------|----------------|----------|--------|----------|-----------|-----------|---------------------------------------------------|
+| id          | Unique key for the Inline Hook.                           | String         | FALSE    | TRUE   | TRUE     |           |           |                                                   |
+| status      | Status of the Inline Hook, INACTIVE will block execution. | Enum           | FALSE    | FALSE  | FALSE    |           |           | One of ACTIVE, INACTIVE                           |
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Re-worked table to avoid having empty cells.

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md [hunk 25]
```diff
@@ -0,0 +1,318 @@
...
+
+| Property | Description                        | Data Type                    |
+|----------|------------------------------------|------------------------------|
+| claims   | Claims included in the token.      | [claims](#claims) object     |
+| lifetime | Lifetime of the token.             | [lifetime](#lifetime) object |
+| scopes   | The scopes contained in the token. | [scopes](#scopes) object     |
+
+#### claims
+
+Consists of name-value pairs for each included claim. See [Tokens and Claims](/docs/api/resources/oidc#tokens-and-claims), in the documentation for Okta's OpenID Connect and OAuth 2.0 API, for descriptions of the claims that can be included.
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: Changed to suggested wording.

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md [hunk 26]
```diff
@@ -0,0 +1,318 @@
...
+
+#### lifetime
+
+| Property   | Description                                                            | Data Type |
+|------------|------------------------------------------------------------------------|-----------|
+| expiration | The expiration time of the token in seconds since January 1, 1970 UTC. | Number    |
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: Thanks, Yu. Corrected to say: "Time in seconds until the token expires."

[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: Yu Liu pointed out in a comment that I got this wrong: it is actually a relative time. So, have changed the description to say: "Time in seconds until the token expires."

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md (hunk 16 above)
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: Changed to use suggested wording.

File: _source/_use_cases/inline_hooks/index.md [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+
+## What Are Okta Inline Hooks?
+
+Inline hooks are outbound calls from Okta to your own custom code, triggered at specific points in Okta process flows. They allow you to integrate custom functionality into those flows.
+
+You implement your custom code as a web service with an Internet-accessible endpoint. It's your responsibility to arrange hosting of your code on a system external to Okta. Okta defines the REST API contract for the requests it sends to your custom code, as well as for the responses your custom code can send back. 
+
+The outbound call from Okta is called a hook. Your code, which receives the call, is referred to as your external service.
+
+The Okta process that triggered the hook is paused until a response from your service is received (i.e., inline hooks are synchronous calls).
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/index.md]: Changed to suggested wording.

File: _source/_use_cases/inline_hooks/index.md [hunk 28]
```diff
@@ -0,0 +1,127 @@
...
+
+## Inline Hook Process Flow
+
+### Extension Points
+
+The points in Okta process flows where inline hooks can be triggered are called extension points, because they are where you can extend Okta functionality. Each type of inline hook is triggered at a particular extension point in a particular Okta process flow. At an extension point, if you have configured an inline hook, Okta calls your external service, and waits for a response. When the response is received, Okta resumes the process flow.
+
+### Inline Hook Call within an Okta Process Flow
+
+The graphic below illustrates the the sequence of steps:
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/index.md]: Corrected.

File: _source/_use_cases/inline_hooks/index.md [hunk 29]
```diff
@@ -0,0 +1,127 @@
...
+Okta uses an HTTPS POST request to call your service.
+
+### HTTP Header
+
+The header of the request sent by Okta includes the following fields:
+
+```http
+Accept: application/json
+Content-Type: application/json
+Authorization: <key>
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/index.md]: Done.

File: _source/_use_cases/inline_hooks/index.md [hunk 30]
```diff
@@ -0,0 +1,127 @@
...
+You need to return an HTTP status code with your response. Typically, your service should return an HTTP status code of 200 (OK). In inline hook types that support empty responses, HTTP status code 204 (No Content) needs to be provided when sending an empty response.
+
+
+#### Don't Use HTTP Status Code to Return Information
+
+Don't use the HTTP status code to return information to Okta regarding problems your service has detected in the data; use an [error](#error) object sent in the JSON payload of the response. HTTP error codes should not be used unless your service could not parse the request from Okta.
+
+### JSON Payload Objects
+
+You can include any of the following types of objects in the JSON payload:
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/index.md]: There are various possibilities: supplying either of those two objects, supplying both of them , supplying none (an empty response).... So, I'd like to keep the existing wording.

File: _source/_use_cases/inline_hooks/index.md [hunk 31]
```diff
@@ -0,0 +1,127 @@
...
+## Inline Hook Setup
+
+After creating your external service, you need to tell Okta it exists, and enable it for a particular process flow. The steps are:
+
+1. Create an external service.
+
+1. Register your service's endpoint with Okta by making a `POST` request to `/api/v1/inlineHooks`. See [Inline Hooks Management API](/docs/api/resources/inline-hooks).
+
+1. Associate the endpoint with a particular Okta process flow. How to do this varies by inline hook type.
+
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/index.md]: Have added the following sentence: "For more information on implementing inline hooks, see the documentation for specific inline hook types linked to in [Currently-Supported Types](#currently-supported-types)."

File: _source/_use_cases/inline_hooks/index.md [hunk 32]
```diff
@@ -0,0 +1,127 @@
...
+### JSON Payload Objects
+
+The JSON payload is where Okta provides specific information about the process flow that's being executed, so that your external service can evaluate the specific situation. Information is encapsulated in JSON objects. The set of objects sent depends on the type of inline hook you're using. Objects are defined in the specific documentation for each type of inline hook.
+
+The objects providing this information are nested within a larger object called `data`.
+
+Always included is `data.context`, providing context information. In general, `data.context` encapsulates Okta objects that your external service cannot affect, while objects in `data` that are outside of `data.context` encapsulate objects that your external service does have the ability to affect, by means of the commands it sends in its response.
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/index.md]: The objects are very big, and are different for different inline hook types, so would rather hold off on this for now.

File: _source/_use_cases/inline_hooks/index.md [hunk 33]
```diff
@@ -0,0 +1,127 @@
...
+
+### JSON Payload Objects
+
+You can include any of the following types of objects in the JSON payload:
+
+#### commands
+
+Lets you return commands to Okta to affect the process flow being executed and to modify values within Okta objects. The available commands differ by inline hook type and are defined in the specific documentation for each inline hook type.
+
+The `commands` object is an array, allowing you to return more than one command in your response. Each element within the array needs to consist of a pair of `type` and `value` elements. Each `type` element needs to be the name of a supported command you wish to invoke. The corresponding `value` element is the operand you wish to specify for the command.
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/index.md]: Because this varies widely between hook types, this could get confusing. I would rather hold off for now.

File: _source/_use_cases/inline_hooks/index.md [hunk 34]
```diff
@@ -0,0 +1,127 @@
...
+
+The names of commands follow Java-style reverse DNS name format, beginning with `com.okta`, followed by an Okta object that the command operates on, and then an action.
+
+#### error
+
+Lets you return error messages. How the error data is used varies by inline hook type.
+
+Within an `error` object, you need to provide an `errorSummary` property set to a text string. Additionally, you can use an `errorCauses` object to supply more information. A single error object can contain multiple `errorCauses` objects. The fields within errorCauses are: `errorSummary`, `reason`, `locationType`, `location`, and `domain`.
```
[the remark this answers is not in the document]
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/index.md]: Again, varies between hook types. We can re-visit, but would like to hold off for now.

File: _source/_use_cases/inline_hooks/index.md [hunk 35]
```diff
@@ -0,0 +1,128 @@
...
+
+## What Are Okta Inline Hooks?
+
+Inline hooks are outbound calls from Okta to your own custom code, triggered at specific points in Okta process flows. They allow you to integrate custom functionality into those flows.
+
+You implement your custom code as a web service with an Internet-accessible endpoint. It's your responsibility to arrange hosting of your code on a system external to Okta. Okta defines the REST API contract for the requests it sends to your custom code, as well as for the responses your custom code can send back. 
+
+The outbound call from Okta is called a hook. Your code, which receives the call, is referred to as your external service.
+
+Inline hooks use asynchronous calls, which means that the Okta process that triggered the hook is paused until a response from your service is received.
```
Comment (amoghvasekar-okta) [inline _source/_use_cases/inline_hooks/index.md]: Aren't they actually `synchronous`? Is this a typo or intended?

File: _source/_docs/api/resources/inline-hooks.md [hunk 36]
```diff
@@ -0,0 +1,578 @@
...
+| id          | Unique key for the Inline Hook.                             | String         | FALSE    | TRUE   | TRUE     | Assigned                                          |
+| status      | Status of the Inline Hook. `INACTIVE` will block execution. | String         | FALSE    | FALSE  | FALSE    | Must be either `ACTIVE` or `INACTIVE`.            |
+| name        | Display name for Inline Hook.                               | String         | FALSE    | TRUE   | FALSE    | Must be between 1 and 255 characters in length.   |
+| type        | Type of the Inline Hook.                                    | inlineHookType | FALSE    | FALSE  | TRUE     | Immutable after Inline Hook creation.             |
+| version     | Version of the Channel.                                     | Integer        | FALSE    | FALSE  | TRUE     | Must match a valid version number.                |
+| channel     | Channel for the Inline Hook.                                | Channel        | FALSE    | FALSE  | FALSE    | Validation is determined by the specific Channel. |
```
Comment (amoghvasekar-okta) [inline _source/_docs/api/resources/inline-hooks.md]: I think my comment got lost. Is the `channel` object model elaborated somewhere, especially things like `headers` and `authScheme` ?

File: _source/_use_cases/inline_hooks/index.md (hunk 35 above)
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/index.md]: Thanks for catching that, @amoghvasekar-okta. Have now corrected it.

File: _source/_docs/api/resources/inline-hooks.md (hunk 36 above)
Comment (motikagan-okta) [reply _source/_docs/api/resources/inline-hooks.md]: I'm still working on your comments @amoghvasekar-okta. For this one, I think I might need to hold off, and create a Jira item for it, to add this later, because I don't have the information.

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md (hunk 15 above)
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: @yuliu-okta, how to document the `data.context` objects is a big question. I don't know right now, so think I need to hold off, and then follow up later. I need clarification on how consistent the objects will be across different hook types, whether context objects with the same name will be identical across different hook types. Also, which objects are likely to be useful to developers.

File: _source/_docs/api/resources/inline-hooks.md (hunk 36 above)
Comment (amoghvasekar-okta) [reply _source/_docs/api/resources/inline-hooks.md]: Sounds good, thanks. Lemme know how I can help or when you want me to give this PR another look.

File: _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md (hunk 13 above)
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/api_am_hook/api_am_hook.md]: Have added the following note: "The `commands` array should only contain commands that can be applied to the requested tokens. For example, if the token is an ID token, the `commands` array should not contain commands of the type `com.okta.access.patch`."

File: _source/_use_cases/inline_hooks/index.md (hunk 5 above)
Comment (motikagan-okta) [reply _source/_use_cases/inline_hooks/index.md]: Amogh, I think I need to to do this in a separate pull request later, just because of the release deadline. I will follow up with you.
